### PR TITLE
Fix all clippy warnings and add clippy to the CI

### PR DESCRIPTION
**Pull request by koute**
Clippy doesn't like the theoretically redundant closure calls that I've added in my last PR, and since clippy doesn't run on the CI here I didn't notice.

While I was at it I also fixed all other clippy warnings, plus one error as it was complaining about an `unsafe` that I've now removed. I've also added clippy to the CI (hopefully it works).

Fixes https://github.com/paritytech/parity-scale-codec/issues/454

Fixes https://github.com/paritytech/parity-scale-codec/issues/456

### CODE REVIEW

**Review by ggwpez — COMMENTED** [review]
Maybe we should add a Clippy config file?  
Could be the one from Substrate in `.cargo/config.toml`.

File: CHANGELOG.md
Comment (ggwpez) [inline CHANGELOG.md:12]: And the unneeded return warning?

File: src/codec.rs
Comment (ggwpez) [inline src/codec.rs:1097]: Not sure if saturating add would make sense here.

File: src/compact.rs
Comment (ggwpez) [inline src/compact.rs:496]:
```suggestion
			3 => {
```
:sparkles:

File: src/encode_append.rs
Comment (ggwpez) [inline src/encode_append.rs:90]: Nit: I think the work `append` is also used in the `sp-io` storage API to refer to this operation.

File: src/encode_append.rs
Comment (ggwpez) [inline src/encode_append.rs:123]: I dont quite understand the code. Why is it using `mem::size_of` instead of some `Encode::size_hint` or something?  
Or is there some assumed relation between in-memory and encoded size?

**Review by ggwpez — COMMENTED** [review]
Looks good, just some Qs.

**Comment by koute** [comment]
> Maybe we should add a Clippy config file? Could be the one from Substrate in `.cargo/config.toml`.

Hmmm.... I think we should be fine with the default one here, as this is orders of magnitude smaller so we don't need to disable a bunch of lints like in Substrate.

**Comment by ggwpez** [comment]
> Hmmm.... I think we should be fine with the default one here, as this is orders of magnitude smaller so we don't need to disable a bunch of lints like in Substrate.

Its not just disabling, it is also denying some. But only `correctness` and `complexity` as it seems.  
Up to you :shrug:

File: src/encode_append.rs
Comment (koute) [reply src/encode_append.rs:90]: In `std` it's called `extend`, so that's why I called it like that. (:

But I don't particularly care. Do you want me to rename it?

**Review by ggwpez — APPROVED** [review]

File: src/encode_append.rs
Comment (koute) [reply src/encode_append.rs:123]: Ah, you're right! Good catch! It's quite late here so I didn't notice that and just copied what the code was originally doing here, but this *is* incorrect.

In fact, now this is harmless since it's only used for `with_capacity` (as I removed the `unsafe`, but previously with the `unsafe` it looks like we had a potential out-of-bounds write...  yikes.

File: src/encode_append.rs
Comment (koute) [reply src/encode_append.rs]: Okay, I renamed it to be consistent with the method name in the trait where it's used.

File: CHANGELOG.md
Comment (koute) [reply CHANGELOG.md]: Yes, no warnings should appear now. (Or at least they don't appear now when I tested it running `cargo clippy` in `polkadot` :P)

File: src/encode_append.rs
Comment (koute) [reply src/encode_append.rs]: I've simplified this. Not sure if there's a better way of doing this. (We can't really get the exact encoded bytesize of the new items since we haven't iterated over them yet.)

File: src/codec.rs
Comment (koute) [reply src/codec.rs:1097]: Hmm... perhaps, but then it's just a size hint so shouldn't really matter if it overflows.

Comment (ggwpez) [reply src/codec.rs:1097]: I dont know if the runtime is normally compiled with overflow checks enabled (that would be my only concern).

Comment (bkchr) [reply src/codec.rs:1097]: Overflow checks are not enabled for release builds as they would be too slow.

File: src/encode_append.rs
Comment (bkchr) [inline src/encode_append.rs:122]: Here we should use `checked_mul`

**Review by bkchr — APPROVED** [review]

File: CHANGELOG.md
Comment (davxy) [inline CHANGELOG.md:8]: Should we add the date?

**Review by davxy — APPROVED** [review]

File: CHANGELOG.md
Comment (koute) [reply CHANGELOG.md:8]: Yes, I'll do that before I'll merge it (so that the date will actually correspond to the day it'll be released :P)

**Comment by koute** [comment]
Added a fix + test for https://github.com/paritytech/parity-scale-codec/issues/456 and also fixed clippy lints in the derive crate (this was as far as I can see necessary to test the fix for #456). Should be good to go once the CI passes.

File: .gitlab-ci.yml
Comment (ggwpez) [inline .gitlab-ci.yml:51]: Can you not use something like `--workspace --tests`?

Comment (koute) [reply .gitlab-ci.yml:51]: That would run clippy on all of the integration tests, which at this point are not clippy clean, and at this point I didn't feel like also cleaning up all of the tests.

File: tests/clippy.rs
Comment (koute) [reply tests/clippy.rs:24]: Yes, good idea. I'll create an issue.

**Pull request merged**
